src/ring_theory/dedekind_domain.lean

```diff
@@ -692,3 +699,195 @@ theorem ideal.prime_iff_is_prime {P : ideal A} (hP : P ≠ ⊥) :
 ⟨ideal.is_prime_of_prime, ideal.prime_of_is_prime hP⟩
 
 end is_dedekind_domain
+
+section is_integral_closure
+
+/-! ### `is_integral_closure` section
+
+We show that an integral closure of a Dedekind domain in a finite separable
+field extension is again a Dedekind domain. This implies the ring of integers
+of a number field is a Dedekind domain. -/
+
+open algebra
+open_locale big_operators
+
+variables {A K} [algebra A K] [is_fraction_ring A K]
+variables {L : Type*} [field L] (C : Type*) [integral_domain C]
+variables [algebra K L] [finite_dimensional K L] [algebra A L] [is_scalar_tower A K L]
+variables [algebra C L] [is_integral_closure C A L] [algebra A C] [is_scalar_tower A C L]
+
+lemma is_integral_closure.range_le_span_dual_basis [is_separable K L]
+  {ι : Type*} [fintype ι] [decidable_eq ι] (b : basis ι K L)
+  (hb_int : ∀ i, is_integral A (b i)) [is_integrally_closed A] :
+  ((algebra.linear_map C L).restrict_scalars A).range ≤
+    submodule.span A (set.range $ (trace_form K L).dual_basis (trace_form_nondegenerate K L) b) :=
+begin
+  let db := (trace_form K L).dual_basis (trace_form_nondegenerate K L) b,
+  rintros _ ⟨x, rfl⟩,
+  simp only [linear_map.coe_restrict_scalars_eq_coe, algebra.linear_map_apply],
+  have hx : is_integral A (algebra_map C L x) :=
+    (is_integral_closure.is_integral A L x).algebra_map,
+  suffices : ∃ (c : ι → A), algebra_map C L x = ∑ i, c i • db i,
+  { obtain ⟨c, x_eq⟩ := this,
+    rw x_eq,
+    refine submodule.sum_mem _ (λ i _, submodule.smul_mem _ _ (submodule.subset_span _)),
+    rw set.mem_range,
+    exact ⟨i, rfl⟩ },
+  suffices : ∃ (c : ι → K), ((∀ i, is_integral A (c i)) ∧ algebra_map C L x = ∑ i, c i • db i),
+  { obtain ⟨c, hc, hx⟩ := this,
+    have hc' : ∀ i, is_localization.is_integer A (c i) :=
+      λ i, is_integrally_closed.is_integral_iff.mp (hc i),
+    use λ i, classical.some (hc' i),
+    refine hx.trans (finset.sum_congr rfl (λ i _, _)),
+    conv_lhs { rw [← classical.some_spec (hc' i)] },
+    rw [← is_scalar_tower.algebra_map_smul K (classical.some (hc' i)) (db i)] },
+  refine ⟨λ i, db.repr (algebra_map C L x) i, (λ i, _), (db.sum_repr _).symm⟩,
+  rw bilin_form.dual_basis_repr_apply,
+  exact is_integral_trace (is_integral_mul hx (hb_int i))
+end
+
+lemma integral_closure_le_span_dual_basis [is_separable K L]
```

Comment:
Is it worth it to keep the `integral_closure` versions of these results? This is really the only one that does something more interesting than directly call the `is_integral_closure` version.

Reply:
I don't really care. Whatever you prefer.

Reply:
Since removing them is more work, let's just keep them then :)